mscore/resourceManager.cpp

```diff
@@ -373,6 +970,232 @@ void ResourceManager::downloadExtension()
+      return localPath;
+      }
+
+//---------------------------------------------------------
+//   downloadInstallPlugin
+//---------------------------------------------------------
+
+void ResourceManager::downloadInstallPlugin()
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      button->setEnabled(false);
+      button->setText(tr("Analyzing..."));
+      QString page_url = button->property("page_url").toString();
+      PluginPackageDescription new_package;
+      new_package.package_name = button->property("name").toString();
+      analyzePluginPage("https://musescore.org" + page_url, new_package);
+      if (new_package.source == UNKNOWN) {
+            button->setText("Unknown plugin source.");
+            button->setEnabled(true);
+            return;
+            }
+      button->setText(tr("Downloading..."));
+      QString localPath = downloadPluginPackage(new_package, page_url);
+
+      if (installPluginPackage(localPath, new_package)) {
+            // add the new description item to the map
+            pluginDescriptionMap[page_url] = new_package;
+            refreshPluginButton(button->property("row").toInt());
+            writePluginPackages();
+            displayPlugins();
+            }
+      else {
+            button->setText("Bad archive. Try again");
+            button->setEnabled(true);
+            return;
+            }
+      }
+
+void ResourceManager::uninstallPluginPackage() 
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      const QString& url = button->property("page_url").toString();
+      if (!pluginDescriptionMap.contains(url))
+            return;
+      PluginPackageDescription& desc = pluginDescriptionMap[url];
+      // remove qml files first(which may be outside the folder of this plugin)
+      for (QString& path : desc.qml_paths)
+            QFile::remove(path);
+      // then remove the folder
+      QDir d(desc.dir);
+      if (!d.removeRecursively()) {
+            button->setText("Failed, try again");
+            return;
+            }
+      pluginDescriptionMap.remove(url);
+      refreshPluginButton(button->property("row").toInt());
+      writePluginPackages();
+      displayPlugins();
+      }
+
+
+void ResourceManager::refreshPluginButton(int row, bool updated/* = true*/)
+      {
+      // TODO: get button, then plugin url, then update status
+      // install button
+      QPushButton* install = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 3)));
+      install->disconnect();
+      QPushButton* uninstall = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 4)));
+      const QString& page_url = install->property("page_url").toString();
+      bool installed = pluginDescriptionMap.contains(page_url);
+      uninstall->setEnabled(installed);
+      if (installed) {
+            if (updated) {
+                  install->setText("Updated");
```

Comment:
and here

Reply:
Thanks. Fixed in https://github.com/musescore/MuseScore/pull/4833/commits/1d26ad066f4fdc3c5897d2e6c467a6de02ce32f4.